### ISSUE TITLE
Analyze PyTorch's execution architecture for Tabgrad

### ISSUE THREAD

**Issue by isaacperez** [isaacperez/tabgrad#13]
## Problem

Tabgrad aims to support ordinary PyTorch frontend behavior without embedding
the official PyTorch runtime. The project needs evidence about which parts of
PyTorch's execution architecture preserve observable semantics and which parts
exist because of native, historical, distributed, or deployment constraints
that a browser-native implementation should not inherit.

## Why it matters

A shallow imitation of operation names could miss dispatcher behavior, tensor
metadata, views, aliasing, mutation, version tracking, autograd recording,
saved values, error timing, or dynamic eager execution. Copying PyTorch's full
internal structure would instead impose complexity and assumptions that do not
fit Pyodide, JavaScript, CPU, or WebGPU.

## Expected result

Produce a source-backed report that traces representative PyTorch operations
through dispatch, tensor state, forward execution, autograd, and optional
compiled paths. Separate semantic requirements that Tabgrad must preserve from
implementation mechanisms it can simplify or omit, and state the consequences
for every central Tabgrad alternative in parent #11.

## Included and excluded work

Included:

- Current official PyTorch source and documentation for operator schemas and
  registration, dispatcher behavior, tensor and storage state, views and
  aliasing, in-place mutation and versioning, autograd nodes and saved values,
  eager execution, and the relationship of compilation or functionalization
  to eager semantics.
- Representative traces for ordinary forward and backward operations,
  including error and mutation behavior relevant to compatibility.
- Identification of native, distributed, accelerator, and historical
  responsibilities that Tabgrad need not reproduce internally.

Excluded:

- Reimplementing PyTorch, running an exhaustive operator survey, or deciding
  Tabgrad's architecture independently of parent #11.
- Treating `torch.compile`, TorchInductor, or another optional subsystem as a
  required browser design without evidence.
- Production code, dependencies, compatibility claims, or benchmarks outside
  evidence needed by the parent decision.

Expand only when another PyTorch component demonstrably owns a semantic or
control-flow property needed to evaluate the parent alternatives.

## Completion conditions

- [ ] The report pins the examined PyTorch revision and documentation versions
  and cites exact primary-source paths or pages for material claims.
- [ ] It diagrams or explains the control and data flow of representative eager
  forward operations and backward propagation.
- [ ] It identifies how schemas, dispatch keys, tensor metadata, storage,
  views, aliasing, mutation, versioning, autograd recording, saved values,
  devices, data types, and errors affect observable behavior.
- [ ] It explains which compiled or graph transformations are optional layers
  over eager behavior and which semantic constraints they preserve.
- [ ] It separates behavior Tabgrad must emulate from C++, native extension,
  distributed, hardware, and legacy mechanisms that need not be copied.
- [ ] It maps findings to the direct eager, lazy intermediate representation,
  hybrid, and thin tensor-centric alternatives in parent #11.
- [ ] It records contrary evidence, unresolved questions, license implications,
  and experiments that could materially change the parent decision.

## Dependencies and related work

Parent: #11. This report is one independently completable part of the evidence
required by that architectural decision and does not choose the final design.

The complete repository issue and pull-request enumeration performed for #11
found no duplicate PyTorch architecture investigation. Completed #1 and PR #3
define the project compatibility boundary but do not contain this analysis.

## Effects and risks

This research can affect later Python API, tensor, autograd, compatibility,
performance, and architecture decisions. It changes no runtime behavior. The
main risks are mistaking one internal implementation for a public requirement,
reading a single subsystem out of context, or using outdated source. Every
claim must identify its revision and whether it is observed behavior, internal
mechanism, or inference for Tabgrad.

## Exact research question

Which responsibilities and invariants in PyTorch's dispatcher, tensor state,
eager execution, autograd, and optional compilation paths are necessary to
support the intended PyTorch-compatible frontend, and which implementation
structures can Tabgrad omit or replace in a browser-native runtime?

## Decision this research will inform

Parent #11 will use the report to decide the central execution abstraction and
where Tabgrad must place operator dispatch, tensor state, realization,
autograd, scheduling, and backend boundaries without precluding required
PyTorch behavior.

## Alternatives to compare

- Preserve PyTorch-like eager dispatch and autograd responsibilities through a
  simplified central engine and operator registry.
- Preserve semantics through a lazy graph or intermediate representation with
  explicit realization.
- Use an eager-compatible hybrid that records or transforms internal regions
  without exposing graph execution to Python callers.
- Use thinner tensor and backend calls only where evidence shows that no
  central dispatcher or graph responsibility is required.

The report evaluates how each alternative can preserve the observed invariants;
it does not select one.

## Evidence already available

- [PyTorch source](https://github.com/pytorch/pytorch) and
  [PyTorch developer documentation](https://pytorch.org/docs/stable/).
- Tabgrad's `README.md` defines the intended compatibility and browser-runtime
  boundary.
- Parent #11 lists the initial alternatives and representative behaviors.

These sources establish where to investigate, not which design Tabgrad should
choose.

## Proposed experiment or investigation

1. Pin a current PyTorch revision and authoritative documentation versions.
2. Locate the owners of operator schema and registration, dispatch selection,
   tensor and storage metadata, views and mutation, autograd recording and
   execution, and compiled graph entry points.
3. Trace a small set of representative operations through these owners: an
   elementwise operation, a view followed by mutation, a reduction, matrix
   multiplication plus activation, a branching backward graph, and a
   data-dependent eager decision.
4. For each observed mechanism, state the user-visible invariant, whether it is
   fundamental or implementation-specific, and which parent alternatives can
   preserve it.
5. Challenge the report for missing ownership, optional-layer confusion, and
   conclusions broader than the inspected paths.

## Environments, inputs, and measurements

Record the PyTorch commit, documentation version, inspected paths, commands,
and any executable trace used. When runtime observation is needed, record the
Python and PyTorch versions, platform, input shapes and data types, device,
expected values or errors, gradients, alias relationships, and mutation state.

This issue does not compare browser performance. Any measurement must only
clarify PyTorch behavior and must not be compared with another architecture
using different workloads or environments.

## Required research record

Record the actual method, source revisions, operation traces, responsibility
map, semantic invariants, mechanisms Tabgrad can omit, implications for each
parent alternative, limitations, contrary evidence, unresolved questions, and
reproducible commands in this issue. Link every material source and clearly
separate observation from inference.

**Comment by isaacperez** [comment]
## Initial source report — PyTorch execution architecture

This is the bounded source report requested by #13 for parent #11. It records evidence and implications; it is **not** a final Tabgrad architecture choice or approval.

### Scope and reproducible method

- Tabgrad: clean `main` at `43339ff2d563b6de68c456bd8d1d98a204b45938`.
- PyTorch source: `main` at [`3cc7a9838ff3a501a54046254ce8fec67bd54ec1`](https://github.com/pytorch/pytorch/tree/3cc7a9838ff3a501a54046254ce8fec67bd54ec1), queried 2026-09-04.
- Stable PyTorch documentation redirected to 2.14; tag `v2.14.0` resolved to `2b3ec34829036a65cd9d1398ea72a0167dc37470`.
- The prepared Tabgrad Python 3.11.5 environment did not contain `torch`; no dependency was installed. The traces below are source/document traces, not runtime-oracle results.
- Method: complete issue/rule inspection; `git ls-remote`; bounded searches and numbered inspection of the pinned primary source and 2.14 documentation. No repository, Git, GitHub, or external-service state was changed during the delegated investigation.
- License: the pinned PyTorch source uses permissive three-condition BSD terms requiring notices and prohibiting endorsement ([LICENSE](https://github.com/pytorch/pytorch/blob/3cc7a9838ff3a501a54046254ce8fec67bd54ec1/LICENSE)). No code was copied.

### Observed ownership and flow

```text
Python torch/Tensor call
  -> generated binding parses overload and converts arguments
  -> ATen schema records signature plus alias/mutation annotations
  -> dispatcher combines tensor and active-mode keys and selects a kernel
  -> Autograd / ADInplaceOrView / functionalization wrappers may act and redispatch
  -> meta, composite, or backend implementation validates and executes
  -> TensorImpl carries logical metadata; StorageImpl owns backing storage

backward()
  -> engine dependency counts and ready queues
  -> Node::apply after incoming gradients are ready
  -> saved values are unpacked and version-checked
  -> gradients are validated, combined, propagated, and accumulated
```

Primary evidence:

- Generated Python bindings parse signatures and call typed ATen functions; code generation and the C++ calling convention are mechanisms, not public semantics ([binding model](https://github.com/pytorch/pytorch/blob/3cc7a9838ff3a501a54046254ce8fec67bd54ec1/torchgen/api/python.py#L33-L90), [generation template](https://github.com/pytorch/pytorch/blob/3cc7a9838ff3a501a54046254ce8fec67bd54ec1/tools/autograd/gen_python_functions.py#L834-L884)).
- Schemas distinguish functional, mutating, and aliasing operators, including `add`, `add_`, and `view` ([schemas](https://github.com/pytorch/pytorch/blob/3cc7a9838ff3a501a54046254ce8fec67bd54ec1/aten/src/ATen/native/native_functions.yaml#L541-L573), [view](https://github.com/pytorch/pytorch/blob/3cc7a9838ff3a501a54046254ce8fec67bd54ec1/aten/src/ATen/native/native_functions.yaml#L8189-L8208)).
- Dispatch extracts and combines argument keys with active modes, then uses an operator-owned registration table ([extractor](https://github.com/pytorch/pytorch/blob/3cc7a9838ff3a501a54046254ce8fec67bd54ec1/aten/src/ATen/core/dispatch/DispatchKeyExtractor.h#L16-L47), [dispatch](https://github.com/pytorch/pytorch/blob/3cc7a9838ff3a501a54046254ce8fec67bd54ec1/aten/src/ATen/core/dispatch/Dispatcher.h#L782-L903)). The global singleton, boxed stack, RAII registration, thread-local machinery, and full key taxonomy are PyTorch-specific mechanisms.
- `TensorImpl` separates sizes, strides, offset, dtype, device, and dispatch state from `StorageImpl`, enabling views to share data with different geometry ([TensorImpl](https://github.com/pytorch/pytorch/blob/3cc7a9838ff3a501a54046254ce8fec67bd54ec1/c10/core/TensorImpl.h#L439-L455), [StorageImpl](https://github.com/pytorch/pytorch/blob/3cc7a9838ff3a501a54046254ce8fec67bd54ec1/c10/core/StorageImpl.h#L32-L68)).
- Every normal tensor has a version counter; views, detach, and unpacked saved values may share it so mutation that invalidates gradients can be detected ([versioning](https://github.com/pytorch/pytorch/blob/3cc7a9838ff3a501a54046254ce8fec67bd54ec1/c10/core/TensorImpl.h#L288-L417)).
- Generated autograd wrappers determine tracking, reject forbidden in-place cases, create nodes/edges, save only derivative inputs, redispatch below Autograd, attach history, and save outputs ([wrapper order](https://github.com/pytorch/pytorch/blob/3cc7a9838ff3a501a54046254ce8fec67bd54ec1/tools/autograd/gen_variable_type.py#L2268-L2300), [derivative declarations](https://github.com/pytorch/pytorch/blob/3cc7a9838ff3a501a54046254ce8fec67bd54ec1/tools/autograd/derivatives.yaml#L229-L232)).
- Saved variables snapshot and later check the shared version; backward waits for all incoming contributions, combines them, and releases saved values unless retained ([SavedVariable](https://github.com/pytorch/pytorch/blob/3cc7a9838ff3a501a54046254ce8fec67bd54ec1/torch/csrc/autograd/saved_variable.cpp#L14-L225), [engine](https://github.com/pytorch/pytorch/blob/3cc7a9838ff3a501a54046254ce8fec67bd54ec1/torch/csrc/autograd/engine.cpp#L1103-L1468)).

### Representative semantic traces

1. `relu(x @ w + b).sum()` shows that one public operator need not equal one kernel. `matmul` selects or decomposes into rank-specific operations; add validation establishes shape, broadcasting, dtype, allocation, and device rules before the backend loop; autograd records only executed nodes and propagates through sum, ReLU, add, and matmul ([matmul schema](https://github.com/pytorch/pytorch/blob/3cc7a9838ff3a501a54046254ce8fec67bd54ec1/aten/src/ATen/native/native_functions.yaml#L3779-L3789), [implementation](https://github.com/pytorch/pytorch/blob/3cc7a9838ff3a501a54046254ce8fec67bd54ec1/aten/src/ATen/native/LinearAlgebra.cpp#L1971-L2018)). Composite lowering is compatible with eager behavior only if metadata, gradients, aliases, and errors remain equivalent.
2. `view` followed by mutation creates different geometry over shared storage. Tracked in-place operations reject invalid leaf/view mutations, otherwise mutate and increment a shared version; a saved alias may intentionally fail later during backward when unpacked ([view implementation](https://github.com/pytorch/pytorch/blob/3cc7a9838ff3a501a54046254ce8fec67bd54ec1/aten/src/ATen/native/TensorShape.cpp#L1912-L1940), [in-place checks](https://github.com/pytorch/pytorch/blob/3cc7a9838ff3a501a54046254ce8fec67bd54ec1/torch/csrc/autograd/VariableTypeUtils.h#L33-L87)). Sharing a buffer without alias/effect identity and history repair is insufficient.
3. In a branching graph, dependency counts delay a predecessor until all gradient contributions arrive. A node DAG, tape, reverse IR, or closures may replace PyTorch's engine, but contributions, requested/unused edges, leaf accumulation, and saved-state retention must match.
4. A Python condition using `x.sum().item()` observes a realized scalar before choosing a branch; only the chosen path enters the dynamic autograd graph ([autograd mechanics](https://docs.pytorch.org/docs/2.14/notes/autograd.html#how-autograd-encodes-the-history)). Tabgrad may delay kernels, but any Python-visible data, state, or error must be current at its observation point.

### Required semantics versus replaceable machinery

The evidence supports these Tabgrad obligations without requiring PyTorch's implementation shape:

- one authoritative operation contract for overloads, metadata, mutation/alias effects, dtype/device rules, gradients, and unsupported cases;
- deterministic ownership of frontend override, autograd/effects, backend selection, and explicit failure;
- logical tensor/view state distinct from physical storage;
- coherent aliases, shared effect/version identity, and mutation visibility;
- dynamic reverse-mode AD over operations actually executed, including branching contributions and saved-value validity;
- common shape, broadcast, promotion, device, and error rules shared by forward and backward;
- explicit synchronization at Python observation and backend-transfer boundaries.

The global C++ dispatcher, full dispatch-key space, intrusive pointers, native allocators, device threads/streams, extension ABI, distributed mechanisms, YAML/code generation, and named `StorageImpl` boundary can be replaced. PyTorch itself describes storage as inherited historical machinery, contrary to treating that class boundary as semantic.

Functionalization proves some physical mutation can become out-of-place work, but it still tracks aliases/effects and has documented limitations ([2.14 functionalize](https://docs.pytorch.org/docs/2.14/generated/torch.func.functionalize.html)). `torch.compile` is optional and uses graph breaks/guards around eager execution; compiled regions do not establish that a global lazy graph is required ([2.14 compile](https://docs.pytorch.org/docs/2.14/generated/torch.compile.html), [compiler FAQ](https://docs.pytorch.org/docs/2.14/user_guide/torch_compiler/torch.compiler_faq.html)).

### Consequences for #11 alternatives

| Alternative | Evidence-backed consequence |
| --- | --- |
| Direct eager central registry/engine | Best structural fit for immediate metadata/errors, dynamic Python control, versioned mutation, and per-executed-op autograd. It needs a much smaller schema/effect/AD layer, not PyTorch's dispatcher complexity. Cross-op optimization and browser dispatch cost remain open. |
| Lazy graph/IR | Viable only if the IR represents metadata validation, aliases, ordered effects, saved-value lifetimes, and unsupported errors—not just pure values. Scalar reads, Python control, transfers, and some mutation force validation/realization. Functionalization's gaps are contrary evidence against automatic full eager equivalence. |
| Eager-compatible hybrid | PyTorch's optional captured regions prove the shape is viable: preserve an eager semantic shell while optimizing pure regions. Shared validation/gradient rules and explicit guards/realization boundaries are required; dual-mode parity and debugging are costs. Browser benefit is not established here. |
| Thin tensor-centric | Viable only if tensor/runtime modules still provide one coherent operation, effect/version, and AD owner. Direct calls scattered across backend kernels are insufficient; eliminating a large engine object is possible, eliminating central responsibility is not. |

### Limitations and remaining evidence

- `main` source and 2.14 docs were pinned separately but not fully diffed; a `main`-only detail must not become a 2.14 compatibility claim without reconciliation.
- No authorized environment had official PyTorch 2.14 installed, so exact runtime values, error timing/text, `_version`, alias identity, and repeated-backward release were not observed. If these could change the parent choice, use a separately authorized disposable official CPU oracle for the four traces.
- Browser overhead belongs to #12, not this report.
- No independent challenge has yet examined this report; that remains mandatory before any final architectural recommendation.

This initial report is complete enough for the parent comparison and user-priority checkpoint, but #13 remains `In progress`: release/runtime reconciliation, independent challenge, and any material follow-up experiment remain explicit.

**Comment by isaacperez** [comment]
## Completion record

The initial PyTorch source report is now complete for its bounded role in parent
#11.

- It pins source/documentation revisions and license; traces binding, schema,
  dispatch, tensor/storage, mutation/versioning and autograd ownership; separates
  required public semantics from replaceable PyTorch machinery; covers views,
  aliases, mutation, branching autograd and observation; and compares every
  parent alternative with explicit limitations.
- The material semantic questions were exercised by the later gates: #20 fixes
  one authoritative operation-admission contract; #21 fixes pending-effect and
  error responsibilities; #22 tests views, mutation, saved versions, branching,
  VJP/JVP and higher-order composition across candidate packings; #23–#24 define
  finite reusable programs and realization; and #25 executes a bounded real
  training slice through the selected browser runtime/backends.
- The absence of a full PyTorch installation in the initial report is not being
  converted into an unearned exact-compatibility claim. Exact operation-by-
  operation PyTorch compatibility remains future conformance work; it no longer
  changes the central ownership and data/control-flow choice.
- The independent integrated challenge is recorded on
  [#11](https://github.com/isaacperez/tabgrad/issues/11#issuecomment-5562167969).
  It found the PyTorch-derived obligations coherent with #19–#26 and retained the
  limits on compiled training evidence.

This completes #13 as research evidence. It does not independently select or
approve the final Tabgrad architecture.
